### PR TITLE
Add user task metrics to Easy Telemetry

### LINKED ISSUE #2062: On `diagramOpened` or `deployment` collect and send userTask metrics to E.T.
__Is your feature request related to a problem? Please describe.__

For BPMN files, on `diagramOpened` or `deployment` collect and send the following metrics to [E.T.](https://github.com/camunda/camunda-modeler/tree/develop/docs/telemetry):

Metric | Desc | Diagram type
-- | -- | --
tasks.userTask.count | number of tasks of type userTask | BPMN
tasks.userTask.form.count | number of tasks of type userTask with "Form Key" or at least one Form Field defined | BPMN
tasks.userTask.form.embedded | number of tasks of type userTask with "Form Key" starting with "embedded:" defined | BPMN
tasks.userTask.form.generic | number of tasks of type userTask with at least one Form Field defined | BPMN
tasks.userTask.form.external | number of tasks of type userTask with "Form Key" starting with "app:" defined | BPMN
tasks.userTask.form.<tbd> | <new Camunda Forms> | BPMN
tasks.userTask.form.other | number of tasks of type userTask with "Form Key" not starting with "embedded:" or "app:" defined | BPMN

(Table to be finalized, but in principle that is it).

__Why to do this?__

This shall help understand the usefullness of the yet to be introduced CamundaForms feature.

__Describe the solution you'd like__

Concrete we need to do:
- [ ] Implement feature collecting the metrics as defined above (hint: we already collect `processVariablesCount` at the exact same events, so just link in there)
- [ ]  Update https://github.com/camunda/camunda-modeler/tree/develop/docs/telemetry
- [ ]  Inform E.T. team according to [process](https://confluence.camunda.com/display/camBPM/ET+-+Easy+Telemetry)

__Describe alternatives you've considered__

n/a

__Additional context__

### PR DESCRIPTION
**Pull request by azeghers**
This PR bundles all telemetry related logic into a `metrics` folder, which will hopefully be the entry point for future tracking logic.
It implements various user task-related tracking, as defined in https://github.com/camunda/camunda-modeler/issues/2062
On the side, it fixes a slight oversight where any elements in subsequent root elements after the first, ie. multiple participants, would not be taken into account.

Closes #2062

__Acceptance Criteria__

* [ ] Corresponds to the concept 
* [ ] Corresponds to the design 

__Definition of Done__

* [ ] corresponds to [the design principles](https://github.com/bpmn-io/design-principles)
* [ ] corresponds to [the code standards](https://github.com/bpmn-io/bpmn-js/blob/master/.github/CONTRIBUTING.md#creating-a-pull-request)
* [ ] passes Continuous Integration checks
* [ ] available as feature branch on GitHub
  * [ ] contains the cleaned up commit history
  * [ ] commit messages satisfy our [commit message guidelines](https://www.conventionalcommits.org/)
  * [ ] a single commit closes the issue via Closes #issuenr

### CODE REVIEW

**Comment by azeghers** [comment]
PR set to draft until Legal and ET team is notified and has given approval.

**Review by MaxTru — CHANGES_REQUESTED** [review]
https://github.com/camunda/camunda-modeler/tree/develop/docs/telemetry requires an update as well

File: client/src/util/metrics/metrics.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * Camunda licenses this file to you under the MIT; you may not use this file
+ * except in compliance with the MIT License.
+ */
+
+import { extractProcessVariables } from './processVariables';
+import { getUserTaskMetrics } from './userTasks';
+
+export default async function(file) {
+  const metrics = {
```
Comment (pinussilvestrus) [inline client/src/util/metrics/metrics.js:15]: personally, I wouldn't bootstrap the metrics here to this extend. We already know that we will have more and more metrics in the future, therefore I would keep the initialization as short as possible

```js
let metrics = { }
```

and let the metrics sub-modules do the trick

```js
const userTaskMetrics = await getUserTaskMetrics(file);

metrics = {
  ...metrics,
  tasks: {
    userTask: userTaskMetrics
  }
};
```

File: client/src/util/metrics/userTasks.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * distributed with this work for additional information regarding copyright
+ * ownership.
+ *
+ * Camunda licenses this file to you under the MIT; you may not use this file
+ * except in compliance with the MIT License.
+ */
+
+import { getFormFields, is, parse } from './helpers';
+
+export async function getUserTaskMetrics(file) {
```
Comment (pinussilvestrus) [inline client/src/util/metrics/userTasks.js:13]: `getUserTaskMetrics` sounds like general metrics in the scope of user tasks to me, not only for forms. Therefore  let the `getUserTaskMetrics`  return 

```js
{
  form: { /* all the stuff you are detecting */ },
  // other stuff to come in the future
}
``` 

or simply rename it to `getUserTaskFormMetrics´

File: client/src/util/metrics/processVariables.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import { getProcessVariables } from '@bpmn-io/extract-process-variables';
+import { parse } from './helpers';
+
+export async function extractProcessVariables(file) {
+  const { contents } = file;
+  const processVariables = [];
+  const definitions = await parse(contents);
+  const rootElements = definitions.get('rootElements');
+
+  rootElements.forEach((element) => processVariables.push(...getProcessVariables(element)));
```
Comment (pinussilvestrus) [inline client/src/util/metrics/processVariables.js:20]: I'm personally not a big fan of these single-liners doing multiple stuff. But that's maybe a matter of taste.

File: client/src/util/metrics/helpers.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
+/**
```
Comment (pinussilvestrus) [inline client/src/util/metrics/helpers.js:1]: are these only helpers for `metrics` but maybe a general `ExtensionElementHelper`?

**Review by pinussilvestrus — CHANGES_REQUESTED** [review]
The first round of review done. A few things we should consider, especially in terms of extendability and readability.

File: client/src/util/metrics/userTasks.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+async function getUserTasks(file) {
+  const definitions = await parse(file.contents);
+  const processes = definitions.rootElements.filter((e) => is(e, 'bpmn:Process'));
+
+  const userTasks = [];
+
+  processes.forEach((process) => {
+    process.flowElements && userTasks.push(...process.flowElements.filter((element) => is(element, 'bpmn:UserTask')));
```
[the remark this answers is not in the document]
Comment (pinussilvestrus) [reply client/src/util/metrics/userTasks.js:27]: It's a huge single-liner as well, maybe we can make it more explicit.

**Comment by pinussilvestrus** [comment]
> https://github.com/camunda/camunda-modeler/tree/develop/docs/telemetry requires an update as well

Shall we hook those docs changes into this PR  as well or is a separated PR okay?

File: client/src/util/metrics/helpers.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+ * except in compliance with the MIT License.
+ */
+
+import BpmnModdle from 'bpmn-moddle';
+
+import CamundaBpmnModdle from 'camunda-bpmn-moddle/resources/camunda';
+
+import { filter } from 'min-dash';
+
+export function is(element, type) {
```
Comment (pinussilvestrus) [inline client/src/util/metrics/helpers.js:17]: We can use our existing helpers here

```js
import { is } from 'bpmn-js/lib/util/ModelUtil';
```

**Comment by MaxTru** [comment]
> > https://github.com/camunda/camunda-modeler/tree/develop/docs/telemetry requires an update as well
> 
> Shall we hook those docs changes into this PR as well or is a separated PR okay?

It is the same concern and I would not like to merge this PR here without having updated the docs. So lets add it to this PR. (Our definition of done shall be "is tested and documented") :-)

File: client/src/util/metrics/helpers.js (hunk 4 above)
Comment (azeghers) [reply client/src/util/metrics/helpers.js:1]: Where would you put them?

Comment (pinussilvestrus) [reply client/src/util/metrics/helpers.js:1]: Maybe just one directory up in `client/src/util`, maybe just as `ModelUtil`.

File: client/src/util/metrics/userTasks.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (azeghers) [reply client/src/util/metrics/userTasks.js:27]: > Does this work if we have user tasks inside other container elements?

It doesn't seem to, in its current state. Thanks for finding that.
Is there another case where a user task could be nested?

File: client/src/util/metrics/parse.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  const moddle = new BpmnModdle({
+    camunda: CamundaBpmnModdle,
+  });
+
+  const { rootElement: definitions } = await moddle.fromXML(xml);
+
+  return definitions;
+}
+
+function extractFlowElementsOfType(rootElement, type, flowElements) {
```
Comment (pinussilvestrus) [inline client/src/util/metrics/parse.js:28]: We have [helper functions](https://github.com/bpmn-io/extract-process-variables/blob/master/src/util/ElementsUtil.js#L75) that solve this recursion problem for us. I wonder if we could re-use that.

```js
import { filter } from 'min-dash';

// collect all elements in a flat array
const elements = selfAndAllFlowElements([ rootElement ], false);

return filter(elements, (element) => {
  return is(element, type);
});
```

**Comment by nikku** [comment]
@MaxTru Do you have another look into this one? Let's see if we get it merged.

File: docs/telemetry/README.md [hunk 8]
```diff
@@ -123,11 +135,22 @@ Furthermore, we add the component which initiated the deployment:
...
+In case of BPMN files, we add the number of available process variables in the diagram as part of the diagram metrics,
+as well as relevant information about user tasks and forms:
 
 ```json
 {
   "diagramMetrics": {
-    "processVariablesCount": 5
+    "processVariablesCount": 5,
+    "tasks": {
+      "userTask": {
```
Comment (pinussilvestrus) [inline docs/telemetry/README.md:146]: `userTask.count` is missing here, right?

File: client/src/plugins/usage-statistics/event-handlers/__tests__/DiagramOpenEventHandlerSpec.js [hunk 9]
```diff
@@ -207,6 +213,248 @@ describe('<DiagramOpenEventHandler>', () => {
       diagramType: 'cmmn'
     });
   });
+
```
Comment (pinussilvestrus) [inline client/src/plugins/usage-statistics/event-handlers/__tests__/DiagramOpenEventHandlerSpec.js:216]: I'd personally create tests for the dedicated metrics ( a - process variables, b - user task forms) and then only test for those. Rationale: we shouldn't be forced to touch the tests whenever we add a new metric, only adding new tests for the new type of metrics.

So what I mean is something like

```js
describe('element templates, () => {

  it('should send element templates');


  it('should resend minimal data if payload is too big');

});


describe('diagram metrics - process variables', () => {  

  it('should send');
  

  it('should send for multiple participants');


  it('should send none');

});


describe('diagram metrics - user tasks', () => {

  it('should send - root level user tasks');


  it('should send - pools');


  it('should send - sub processes');


  it('should send - empty');
});
```

And then for a single test, to only test one thing at a time

```js
it('should send - root level user tasks', async () => {
  
// ...
  const metrics = onSend.getCall(0).args[0].diagramMetrics;

  const { userTask } = metrics.tasks;

  // then
  expect(userTask).to.equal(...);
  
});
```

**Review by pinussilvestrus — COMMENTED** [review]
Looks good from the implementation. 👍 

I'd like to update the test structure to make it cleaner.

File: client/src/plugins/usage-statistics/event-handlers/__tests__/DiagramOpenEventHandlerSpec.js (hunk 9 above)
Comment (azeghers) [reply client/src/plugins/usage-statistics/event-handlers/__tests__/DiagramOpenEventHandlerSpec.js:216]: Good point, I'll update that

File: client/src/util/metrics/processVariables.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  const { contents } = file;
+
+  const processVariables = [];
+
+  const definitions = await getDefinitions(contents);
+
+  const rootElements = definitions.get('rootElements');
+
+  rootElements.forEach((element) => {
+    processVariables.push(...getProcessVariables(element));
```
Comment (pinussilvestrus) [inline client/src/util/metrics/processVariables.js:25]: good catch 👏

**Review by pinussilvestrus — COMMENTED** [review]
Code looks clean for me now and works as expected 👍  Good job 👏 

I'd approve it, but would wait for @MaxTru if the data is 100% as intended (because I was not involved in the discussions), cf. https://github.com/camunda/camunda-modeler/pull/2067/files#diff-7681e0231923157e4f6dbb8cc8b6d3c96a2a053c8143a3a8347cb75975ae0ea3R84

According to the issue description, it should be okay 👍

**Review by MaxTru — APPROVED** [review]
Thanks for the contribution (@azeghers ) and also for the detailed review (@pinussilvestrus ). 

I did not look into the code since this was reviewed already but validated that the documented tracking is in line with what was sent to legal. So :+1:

**Pull request merged**